tests/drivers/pwm/pwm_gpio_loopback/src/main.c

```diff
@@ -0,0 +1,303 @@
+			TC_PRINT("PWM output matches the programmed values\n");
+			return TC_PASS;
+		}
+
+		TC_PRINT("PWM output does NOT match the programmed values\n");
+		return TC_FAIL;
+	}
+}
+
+static void test_run(uint32_t period, uint32_t pulse, uint8_t unit, bool inverted)
+{
+	int result;
+	pwm_flags_t flags = inverted ? PWM_POLARITY_INVERTED : 0;
+
+	TC_PRINT("Test case: [Period: %" PRIu32 "] [Pulse: %" PRIu32
+		 "] [Unit: %s] [Inverted: %s]\n",
+		 period, pulse, (unit == UNIT_CYCLES) ? "cycles" : "nsec", inverted ? "Yes" : "No");
+
+	if (unit == UNIT_CYCLES) {
+		result = pwm_set_cycles(ctx.pwm_dev, DEFAULT_PWM_PORT, period, pulse, flags);
+		zassert_false(result, "Failed on pwm_set_cycles() call");
+	} else { /* unit == UNIT_NSECS */
+		result = pwm_set(ctx.pwm_dev, DEFAULT_PWM_PORT, period, pulse, flags);
+		zassert_false(result, "Failed on pwm_set() call");
+	}
+
+	result = check_timing(period, pulse, unit, inverted);
+
+	zassert_equal(result, TC_PASS, "Test case failed");
+}
+
+ZTEST_USER(pwm_gpio_loopback, test_pwm_nsec)
+{
+	/* Test case: [Duty: 25%] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PULSE_NSEC, UNIT_NSECS, false);
+
+	/* Test case: [Duty: 100%] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PERIOD_NSEC, UNIT_NSECS, false);
+
+	/* Test case: [Duty: 0%] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, 0, UNIT_NSECS, false);
+
+	/* Test case: [Duty: 100%+] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PERIOD_NSEC + 10, UNIT_NSECS, false);
+
+	/* Test case: [Duty: 25%] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PULSE_NSEC, UNIT_NSECS, false);
+}
+
+ZTEST_USER(pwm_gpio_loopback, test_pwm_nsec_inverted)
+{
+	/* Test case: [Duty: 25%] [Unit: nsec] [Inverted: Yes] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PULSE_NSEC, UNIT_NSECS, true);
+
+	/* Test case: [Duty: 100%] [Unit: nsec] [Inverted: Yes] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PERIOD_NSEC, UNIT_NSECS, true);
+
+	/* Test case: [Duty: 0%] [Unit: nsec] [Inverted: Yes] */
+	test_run(DEFAULT_PERIOD_NSEC, 0, UNIT_NSECS, true);
+
+	/* Test case: [Duty: 100%+] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PERIOD_NSEC + 10, UNIT_NSECS, true);
+
+	/* Test case: [Duty: 25%] [Unit: nsec] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_NSEC, DEFAULT_PULSE_NSEC, UNIT_NSECS, true);
+}
+
+ZTEST_USER(pwm_gpio_loopback, test_pwm_cycle)
+{
+	/* Test case: [Duty: 50%] [Unit: cycle] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_CYCLE, DEFAULT_PULSE_CYCLE, UNIT_CYCLES, false);
+
+	/* Test case: [Duty: 100%] [Unit: cycle] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_CYCLE, DEFAULT_PERIOD_CYCLE, UNIT_CYCLES, false);
+
+	/* Test case: [Duty: 0%] [Unit: cycle] [Inverted: No] */
+	test_run(DEFAULT_PERIOD_CYCLE, 0, UNIT_CYCLES, false);
+}
+
+static void *pwm_gpio_loopback_setup(void)
+{
+	static struct gpio_callback gpio_cb;
+
+	ctx.pwm_dev = DEVICE_DT_GET(PWM_DEV_NODE);
```

Comment:
A nit here, since we are dealing with a single file here, using a fixture like this really is needed. Just define all the members needed statically at the top of the file.